conjure-go-client/httpclient/authn_test.go

```diff
@@ -113,5 +116,205 @@ func TestRoundTripperWithBasicAuthProvider(t *testing.T) {
+			if assert.Equal(t, bearerAuthHeader, req.Header.Get("Authorization")) {
+				rw.WriteHeader(http.StatusOK)
+			} else {
+				rw.WriteHeader(http.StatusUnauthorized)
+			}
+		}))
+	}
+
+	for _, tc := range []struct {
+		Name          string
+		Server        func(t *testing.T) *httptest.Server
+		Config        httpclient.ClientConfig
+		ClientParams  []httpclient.ClientOrHTTPClientParam
+		RequestParams []httpclient.RequestParam
+	}{
+		{
+			Name:   "NoAuth",
+			Server: noAuthServer,
+		},
+		// Basic Auth
+		{
+			Name:   "BasicAuth config",
+			Server: basicAuthServer,
+			Config: httpclient.ClientConfig{BasicAuth: &httpclient.BasicAuth{User: username, Password: password}},
+		},
+		{
+			Name:         "WithBasicAuth param",
+			Server:       basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuth(username, password)},
+		},
+		{
+			Name:   "WithBasicAuthProvider param",
+			Server: basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthProvider(func(ctx context.Context) (httpclient.BasicAuth, error) {
+				return httpclient.BasicAuth{User: username, Password: password}, nil
+			})},
+		},
+		{
+			Name:   "WithBasicAuthOptionalProvider present",
+			Server: basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthOptionalProvider(func(ctx context.Context) (*httpclient.BasicAuth, error) {
+				return &httpclient.BasicAuth{User: username, Password: password}, nil
+			})},
+		},
+		{
+			Name:   "WithBasicAuthOptionalProvider absent",
+			Server: noAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthOptionalProvider(func(ctx context.Context) (*httpclient.BasicAuth, error) {
+				return nil, nil
+			})},
+		},
+		{
+			Name:         "WithBasicAuth param beats basic config",
+			Server:       basicAuthServer,
+			Config:       httpclient.ClientConfig{BasicAuth: &httpclient.BasicAuth{User: "wrong", Password: "wrong"}},
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuth(username, password)},
+		},
+		{
+			Name:         "WithBasicAuth param beats bearer config",
+			Server:       basicAuthServer,
+			Config:       httpclient.ClientConfig{APIToken: &token},
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuth(username, password)},
+		},
+		// Bearer tokens
+		{
+			Name:   "APIToken config",
+			Server: bearerAuthServer,
+			Config: httpclient.ClientConfig{APIToken: &token},
+		},
+		{
+			Name:   "APITokenFile config",
+			Server: bearerAuthServer,
+			Config: httpclient.ClientConfig{APITokenFile: &apiTokenFile},
+		},
+		{
+			Name:         "WithAuthToken param",
+			Server:       bearerAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithAuthToken(token)},
+		},
+		{
+			Name:         "WithAuthToken param beats basic config",
```

Comment:
this test fails on develop because the header set by WithAuthToken is overwritten by the client's middleware